agent/engine/dependencygraph/graph.go

```diff
@@ -158,6 +163,44 @@ func DependenciesAreResolved(target *apicontainer.Container,
 	return nil, nil
 }
 
+// TaskResourceDependenciesAreResolved validates that the `target` resource can be
+// transitioned given the current known state of the containers in `by`. If
+// this function returns true, `target` should be technically able to transit
+// without issues.
+// Transitions are between known statuses (whether the resource can move to
+// the next known status), not desired statuses; the desired status typically
+// is either CREATED or REMOVED.
+func TaskResourceDependenciesAreResolved(target taskresource.TaskResource,
+	by []*apicontainer.Container) error {
+
+	nameMap := make(map[string]*apicontainer.Container)
+	for _, cont := range by {
+		nameMap[cont.Name] = cont
+	}
+
+	if !verifyContainerDependenciesResolvedForResource(target, nameMap) {
+		return ErrContainerDependencyNotResolvedForResource
+	}
+
+	return nil
+}
+
+func verifyContainerDependenciesResolvedForResource(target taskresource.TaskResource, existingContainers map[string]*apicontainer.Container) bool {
+	targetNext := target.GetKnownStatus() + 1
```

Comment:
i think this should be `target.GetNextKnownStateProgression()` instead. As it, targetNext will be RESOURCE_PROVISION when target is RUNNING, which might not be correct.